src/k2/module/k23si/Module.cpp

```diff
@@ -1108,11 +1108,40 @@ K23SIPartitionModule::_processWrite(dto::K23SIWriteRequest&& request, FastDeadli
         }
     }
 
+    // build write_ranges for Transaction end request
+    std::unordered_map<String, std::unordered_set<dto::KeyRangeVersion>> write_ranges; // will be filled in later if WI is created successfully
+
+    //endRequest created for calling EndTransaction for one hot transactions.
+    dto::K23SITxnEndRequest endRequest{request.pvid,
+        request.trhCollection,
+        request.trh,
+        request.mtr,
+        dto::EndAction::Commit,
+        std::move(write_ranges)
+    };
+
+    return seastar::do_with((dto::K23SITxnEndRequest)std::move(endRequest), (bool)request.isonehot,
+                            [this, &request] (dto::K23SITxnEndRequest& endRequest, bool& isonehot) {
+        // all checks passed - we're ready to place this WI as the latest version
+        auto& vset = _indexer[request.key];
+        auto status = _createWI(std::move(request), vset);
+        K2LOG_D(log::skvsvr, "WI creation with status {}", status);
+
+        // not one hot transaction or WI creation failed
+        if(!isonehot || !status.is2xxOK()) {
+            return RPCResponse(std::move(status), dto::K23SIWriteResponse{});
+        }
```

Comment:
you should create the end request above here only if needed i.e. WI succeeded and this is a 1-shot

Reply:
The problem here was that there is no copy constructor for write request and we need to move the request to create WI. To create an end request, we need the values in the write request, thus we need to create it before the write request is moved. I didn't want to add a copy constructor because it is not needed anywhere else.